drivers/hid/hidclass/fdo.c

```diff
@@ -388,6 +390,215 @@ HidClassFDO_GetDescriptors(
+                                tempIRP->IoStatus.Status = STATUS_INVALID_USER_BUFFER;
+                                tempIRP->IoStatus.Information = 0;
+                            }
+                        }
+
+                        // Mark this IRP to be cenceled after exiting spinLock
+                        InsertHeadList(&completeIRP, &tempIRP->Tail.Overlay.ListEntry);
+                    }
+                }
+            }
+
+            // No IRP waiting data, put it into the buffer
+            if (IsListEmpty(&completeIRP) == TRUE)
+            {
+                HidClass_CyclicBufferPut(&PDODeviceExtension->InputBuffer, DeviceExtension->InputBuffer);
+            }
+
+            KeReleaseSpinLock(&PDODeviceExtension->ReadLock, CurrentIRQL);
+
+            // complete waiting IRPs
+            {
+                PIRP nextIrp;
+                PLIST_ENTRY listEntry;
+                while (IsListEmpty(&completeIRP) == FALSE)
+                {
+                    listEntry = RemoveHeadList(&completeIRP);
+                    nextIrp = CONTAINING_RECORD(listEntry, IRP, Tail.Overlay.ListEntry);
+
+                    IoCompleteRequest(nextIrp, IO_NO_INCREMENT);
+                }
+            }
+        }
+    }
+
+    // re-init read
+    HidClassFDO_SubmitRead(DeviceExtension);
+
+    /* stop completion */
+    return STATUS_MORE_PROCESSING_REQUIRED;
+}
+
+NTSTATUS
+HidClassFDO_SubmitRead(IN PHIDCLASS_FDO_EXTENSION DeviceExtension)
+{
+    PIO_STACK_LOCATION IoStack;
+    NTSTATUS Status;
+
+    /* re-use irp */
+    IoReuseIrp(DeviceExtension->InputIRP, STATUS_SUCCESS);
+
+    /* init irp */
+    DeviceExtension->InputIRP->UserBuffer = DeviceExtension->InputWriteAddress;
+
+    /* get next stack location */
+    IoStack = IoGetNextIrpStackLocation(DeviceExtension->InputIRP);
+
+    /* init stack location */
+    IoStack->MajorFunction = IRP_MJ_INTERNAL_DEVICE_CONTROL;
+    IoStack->Parameters.DeviceIoControl.IoControlCode = IOCTL_HID_READ_REPORT;
+    IoStack->Parameters.DeviceIoControl.OutputBufferLength = DeviceExtension->InputBufferSize;
+    IoStack->Parameters.DeviceIoControl.InputBufferLength = 0;
+    IoStack->Parameters.DeviceIoControl.Type3InputBuffer = NULL;
+    IoStack->DeviceObject = ((PHIDCLASS_PDO_DEVICE_EXTENSION)DeviceExtension->DeviceRelations->Objects[0]->DeviceExtension)->FDODeviceObject;
+    //IoStack->DeviceObject = DeviceExtension->DeviceRelations->Objects[0];
+
+    /* set completion routine */
+    IoSetCompletionRoutine(DeviceExtension->InputIRP, HidClassFDO_ReadCompletion, DeviceExtension, TRUE, TRUE, TRUE);
+
+    // TODO: create function for this.
+    // let's dispatch the request
+    IoSetNextIrpStackLocation(DeviceExtension->InputIRP);
+    Status = DeviceExtension->Common.DriverExtension->MajorFunction[IRP_MJ_INTERNAL_DEVICE_CONTROL](((PHIDCLASS_PDO_DEVICE_EXTENSION)DeviceExtension->DeviceRelations->Objects[0]->DeviceExtension)->FDODeviceObject, DeviceExtension->InputIRP);
```

Comment:
because the HidClass it not driver but some kind of wrapper. Because of those usual methods of calling the uplevel driver not possible here. I'm don't remember exactly what driver will be called in this case. But code here is only way to call usb driver